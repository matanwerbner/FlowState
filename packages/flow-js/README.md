# flow-js

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build flow-js` to build the library.

## Running unit tests

Run `nx test flow-js` to execute the unit tests via [Jest](https://jestjs.io).
