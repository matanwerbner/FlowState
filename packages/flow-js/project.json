{
  "name": "flow-js",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/flow-js/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "build": {
      "executor": "@nx/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/packages/flow-js",
        "main": "packages/flow-js/src/index.ts",
        "tsConfig": "packages/flow-js/tsconfig.lib.json",
        "assets": ["packages/flow-js/*.md"]
      }
    }
  }
}
