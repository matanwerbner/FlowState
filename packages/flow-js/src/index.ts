export * from './lib/flow-js';
