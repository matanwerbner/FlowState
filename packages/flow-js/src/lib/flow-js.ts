import { FlowStateMachine } from './flowStateMachine';
import type { FlowConfig, FlowContext, FlowStates } from './types';

export const createFlow = <
  TPossibleStates extends FlowStates<TPossibleStates, TContext>,
  TContext extends FlowContext
>(
  params: FlowConfig<TContext, TPossibleStates>
) => {
  return new FlowStateMachine<TContext, TPossibleStates>(params);
};
