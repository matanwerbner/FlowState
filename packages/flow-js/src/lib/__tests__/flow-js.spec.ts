import { getTestFlow } from './flow-js.fixture';

describe('flowJs', () => {
  let fixtureFlow: ReturnType<typeof getTestFlow>;

  beforeEach(() => {
    fixtureFlow = getTestFlow();
  });

  it('should create a flow', () => {
    expect(fixtureFlow).toBeDefined();
  });

  it('should return initial state', () => {
    expect(fixtureFlow.currentState).toBe('state1');
  });

  it('should not trigger event if validation fails', async () => {
    await expect(fixtureFlow.triggerEvent('event1')).rejects.toMatchSnapshot();
  });

  it('should return validation errors based on data', async () => {
    await expect(
      fixtureFlow.triggerEvent('event1', { isValid: false })
    ).rejects.toMatchSnapshot();
  });

  it("should fail triggering event not found in current state's transitions", async () => {
    await expect(fixtureFlow.triggerEvent('event2')).rejects.toMatchSnapshot();
  });
});
