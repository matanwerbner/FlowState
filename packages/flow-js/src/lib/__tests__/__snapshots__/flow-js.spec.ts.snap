// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`flowJs should fail triggering event not found in current state's transitions 1`] = `[Error: Event event2 not found in state state1 transitions]`;

exports[`flowJs should not trigger event if validation fails 1`] = `
{
  "errors": {
    "isValid": "context.isValid is not true",
  },
}
`;

exports[`flowJs should return validation errors based on data 1`] = `
{
  "errors": {
    "isValid": "context.isValid is not true",
  },
}
`;
